build_stage_repository

```diff
@@ -172,47 +166,160 @@ def packages_from_comps(comps):
 
-    base_dir = 'tmp'
-    rpm_sync_dir = f"{base_dir}/rpms"
-    srpm_sync_dir = f"{base_dir}/srpms"
 
-    stage_dir = f"{base_dir}/{collection}/{version}/{dist}/"
-    rpm_dir = f"{stage_dir}/{arch}"
-    srpm_dir = f"{stage_dir}/source"
+def prod_repository(collection, version, dist, arch):
+    if collection == "foreman" and version == "nightly":
+        return f"https://yum.theforeman.org/{version}/{dist}/{arch}"
+    else:
+        return f"https://yum.theforeman.org/{collection.replace('foreman-', '')}/{version}/{dist}/{arch}"
 
-    if not os.path.exists(rpm_sync_dir):
-        os.makedirs(rpm_sync_dir)
 
-    if not os.path.exists(srpm_sync_dir):
-        os.makedirs(srpm_sync_dir)
+def copr_repository_urls(repository):
+    cmd = [
+        'dnf',
+        'reposync',
+        '--urls',
+        '--refresh',
+        '--repofrompath',
+        f"copr,{repository}",
+        '--repo',
+        'copr'
+    ]
 
-    if not os.path.exists(rpm_dir):
-        os.makedirs(rpm_dir)
+    urls = check_output(cmd, universal_newlines=True, stderr=STDOUT)
+    return urls.split("\n")
+
+
+def compare_repositories(new_repository, old_repository, source=False):
+    cmd = [
+        'dnf',
+        'repodiff',
+        '--simple',
+        '--refresh',
+        '--compare-arch',
+        '--repofrompath',
+        f"new,{new_repository}",
+        '--repofrompath',
+        f"old,{old_repository}",
+        '--repo-old',
+        'old',
+        '--repo-new',
+        'new'
+    ]
+
+    if not source:
+        cmd.extend(['--arch', 'noarch,x86_64'])
+
+    print(' '.join(cmd))
+    return check_output(cmd, universal_newlines=True)
 
-    if not os.path.exists(srpm_dir):
-        os.makedirs(srpm_dir)
 
-    sync_copr_repository(collection, version, rpm_sync_dir, dist, arch)
-    sync_copr_repository(collection, version, srpm_sync_dir, dist, arch, source=True)
+def parse_repodiff(diff):
+    split_diff = diff.split("\n")
+    packages = []
 
-    move_rpms_from_copr_to_stage(collection, version, srpm_sync_dir, srpm_dir, dist, arch, source=True)
-    move_rpms_from_copr_to_stage(collection, version, rpm_sync_dir, rpm_dir, dist, arch)
+    for line in split_diff:
+        if line.startswith('Added package  :'):
+            packages.append(line.replace('Added package  : ', ''))
 
+        if ' -> ' in line:
+            packages.append(line.split(' -> ')[1])
+
+    return packages
+
+
+def download_copr_packages(packages, urls, repository, downloads_dir, included_packages):
+    for package in packages:
+        name = f"{package}.rpm".replace("-1:", "-")
+
+        for url in urls:
+            if name in url:
+                download_url = url
+
+        if not os.path.exists(f"{downloads_dir}/{name}"):
+            print(f"Downloading {repository}/{name} to {downloads_dir}")
+            urlretrieve(download_url, f"{downloads_dir}/{name}")
```

Comment:
Does `urlretrieve` keep a session open? If it needs to open a new connection for every file it may be a lot slower.

Reply:
I do not know, and some quick searches did not yield any particular result. I've not any particular slowness.

Reply:
I was trying to understand if this maintains a persistent connection between downloads so I went to the docs: https://docs.python.org/3/library/urllib.request.html#urllib.request.urlretrieve

First thing to notice is that it's listed as a legacy method. It also links to `urlopen()` which states:

> urllib.request module uses HTTP/1.1 and includes Connection:close header in its HTTP requests.

So it closes the connection after fetching the file, which means you set up a lot of connections, which has a certain overhead. When I ping the endpoint I have 6ms latency, so it probably isn't that big of a deal. Possible future optimization.